### ISSUE TITLE
PDF header and footer

### ISSUE THREAD

**Issue by mprandot** [barryvdh/laravel-snappy#139]
Hello!

I need to repeat my header and footer on every page, looking at the documentation I realized that I can not include pure HTML in the 'header-html' option (without a doctype).

I did some tests and I can not divide my files (header.blade, content.blade, footer.html), this creates a giant error, can help me or how to proceed in this case?

Thanks! laravel-snappy is amazing!

**Comment by mprandot** [comment]
Another thing, i can load some view in 'header-html' option? 
Like this : 
`$pdf->setOption('header-html', View::make('path.to.header')->render()`

**Comment by barryvdh** [comment]
You need to create a route or html file and pass the full url or path.

**Comment by mprandot** [comment]
Thanks for help!

**Comment by mprandot** [comment]
Not work : 
``` php
    $pdf = App::make('snappy.pdf.wrapper');
    $pdf->setOption('header-html', base_path('views/Modulos/Funcional/OrdemServico/Os/header.blade.php'));
    $pdf->setOption('footer-html', base_path('views/Modulos/Funcional/OrdemServico/Os/footer.blade.php'));
    $pdf->loadView('Modulos.Funcional.OrdemServico.Os.imprime', $parametros);
```
i remove .blade.php and still not work.
Not load my styles in header.blade.php and no repet the header and footer in others pages, my implementation is correct?

**Comment by mprandot** [comment]
@barryvdh, can help me?

**Comment by barryvdh** [comment]
Create a route that renders the html for the header/footer, and pass the full url to that route as option.

**Comment by mprandot** [comment]
Hello @barryvdh, this give me a timeout : 

```log
The process "/usr/local/bin/wkhtmltox/bin/wkhtmltopdf --lowquality --footer-html 'http://localhost:8000/footer' --header-html 'http://localhost:8000/header' '/tmp/knp_snappy581b319fc17132.71963197.html' '/tmp/knp_snappy581b319fc178b4.96324727.pdf'" exceeded the timeout of 60 seconds.
```

```php
$pdf = App::make('snappy.pdf.wrapper');

$pdf->setOption('header-html', 'http://localhost:8000/header');

$pdf->setOption('footer-html','http://localhost:8000/footer');

$pdf->loadView('Modulos.Funcional.OrdemServico.Os.imprime', $parametros);
```

**Comment by spacemudd** [comment]
I had this issue and for me the solution was to inlucde <!DOCTYPE html> in both the header/footer.

**Comment by mprandot** [comment]
@spacemudd, my footer and header has <!DOCTYPE html> :disappointed:

**Comment by mprandot** [comment]
Maybe i've found the problem!

In localhost:8000 the footer and header not work,
i've deployed this code and now with real URL works fine.

Works only with real URL

It's can be the problem? localhost:8000?

**Comment by mprandot** [comment]
Guys!

Found problem!

**Comment by mprandot** [comment]
Guys!

The problem is with localhost:8000. 
I've created a VH to my application and all worked as expected!

**Comment by ArifLateefShah** [comment]
Same procedure Which we are using in the setting header for the first page.

            $header = View::make('reportHeader');
            $pdf = PDF::loadView('report::cashFlowByStages', $projectDetails);
            $pdf->setOption('header-html', $header);
            return $pdf->inline('.Carbon::now()->format('d_m_Y').'documentName.pdf');

**Comment by sleepm** [comment]
@mprandot 
you have to use .html like
'footer-html' => resource_path('views/inware_invoice_footer.html')
but only show one line...
if footer has more than one line will only show one line
pass a url will not work and in command line genreted pdf don't have footer

**Comment by ExCluSiv3** [comment]
@ijpatricio thanks finally something that works :D 

But to "complete" this solution: You **have to** set top and bottom margins to make the header and footer "visible"

**Comment by stuartcusackie** [comment]
I am using a CSS framework so I needed a combination of the answers provided by @ijpatricio and @ExCluSiv3 


`//Prepare and render our PDF header and footer views
        $page_header_html = view()->make('pdf.invoices.partials.page-header', $data)->render();
        $page_footer_html = view()->make('pdf.invoices.partials.page-footer', $data)->render();

        //Set up our options to invluce our header and footer
        //The PDF doesn't render correctly without some of these
        $options = [
            'orientation'   => 'portrait',
            'encoding'      => 'UTF-8',
            'header-html'   => $page_header_html,
            'footer-html'   => $page_footer_html
        ];

        //Create our PDF with the main view and set the options
        $pdf = PDF::loadView('pdf.invoices.default', $data);
        $pdf->setOptions($options);

        //Alternatively, output pdf as file download
        //return $pdf->download('userlist.pdf');

        //return PDF inline
        return $pdf->inline('invoice.pdf');`

**Comment by lucasgollner** [comment]
### This work to me! (I'm using->"barryvdh/laravel-dompdf": "^0.8.3" )
[](https://ourcodeworld.com/articles/read/687/how-to-configure-a-header-and-footer-in-dompdf)

**Comment by lucasgollner** [comment]
This worked to me! (I'm using,  "barryvdh/laravel-dompdf": "^0.8.3" )
https://ourcodeworld.com/articles/read/687/how-to-configure-a-header-and-footer-in-dompdf
[](https://ourcodeworld.com/articles/read/687/how-to-configure-a-header-and-footer-in-dompdf)

**Comment by subdesign** [comment]
What worked for me, adding a header with an image into it:
1. add view to `header-html`
```php 
->setOption('header-html', view('header'))
```
2. add DOCTYPE for the header blade (`views/header.blade.php`)
```php
<!DOCTYPE html>
<img style="width:100%; height:110px" src="http://project.local/images/header-image.jpg">
```
set `width` and `height` too for the image
`img` also worked with `public_path()`

3. set appropriate top margin, to have space for the header (mine was 25)
```php
 ->setOption('margin-top', 25)
```
Nothing less or more needed..

**Comment by moronkids** [comment]
Hello, I have a question, today I successfully use this package. But I have a little problem, I added header then the table in page 2,3,4.... is cropped
![image](https://user-images.githubusercontent.com/38756422/83255621-f98e0c80-a1da-11ea-820a-6e4f53abf381.png)
how to fix this? 
sorry bad English
thank you

**Comment by mmabel84** [comment]
Hi, I'm having an issue with setOption('header-html', $header), whatever I put in '$header', the system crashes with 'The process has been signaled with signal "11".' If I don't use that option, every thing works fine (using footer-html every thing works like a charm). 
I try to use the solution of @ijpatricio but it didn't work, I got this error: 'The option '0' does not exist.' on /vendor/knplabs/knp-snappy/src/Knp/Snappy/AbstractGenerator.php(138). My options array is very simple:
$options = [
            'header-html' => $header,
            'footer-html' => $footer,
            'footer-right', 'Hoja [page] de [toPage]'

        ];
But it seems none of these options are being recognized.
Please, does any one can help me here?
Thanks in advance.

**Comment by ijpatricio** [comment]
Hello @mmabel84 

so this was 2.5 years ago, time flies! naturally, i can't remember everything from top of my head. is the project open source, so i could put my hands on it?

Side note:
also, i would reccomend https://github.com/spatie/browsershot 
it came out later, and based on pupeteer and headless chrome, whatever you want to render, you can, and many may options

**Comment by mmabel84** [comment]
Hi @ijpatricio , sorry for the delay.
It is not open source. The thing is I'm trying to generate a very large document, and only when I use header-html option, it breaks down with the signal 11 error (I have tried passing url, making the header view first and passing the result to the option and nothing seems to work). I thought it was a memory issue, and added ini_set('memory_limit','2048M') in my script, to execute this particular process with more php memory, but it didn't work, the signal 11 is still there.

I gave up with header-html and changed it by header-center and the pdf was generated, but with these options (header-center, header-left, header-right) that receive a text instead html, the utf-8 encoding is not working, the accented marks doesn't appear on the document.

So, at this point, don't know exactly what to do, any help @ijpatricio , @barryvdh ??

So, right now, don't know which way
